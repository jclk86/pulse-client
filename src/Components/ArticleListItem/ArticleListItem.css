.container_article_list_item {
  width: 100%;
  display: flex;
  padding-right: 100px;
  margin-top: 15px;
}

.container_article_preview {
  text-align: center;
  margin: 0 auto;
  border-radius: 3px;
  width: 95%;
  /* display: flex; */
  /* justify-content: space-between; */
  box-shadow: 6px 6px 6px 1px rgb(15, 15, 15);
  background: linear-gradient(rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.1)),
    url("../../images/paper.jpg") no-repeat center center fixed;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  overflow: auto;
  color: black;
}

.container_article_preview:hover {
  border: 1px solid gray;
}

.article_list_item_image {
  width: 100%;
}
/* percentge is preventing size adjusting */
.container_article_list_item_image {
  width: 100%;
  margin-left: 0;
}

.container_article_list_item_info {
  width: 95%;
  padding-bottom: 20px;
}

.article_list_item_title {
  margin-top: 10px;
  display: block;
}

.preview_text {
  display: block;
}

.article_list_item_title_link {
  text-decoration: none;
  color: black;
  font-size: 24px;
}

.container_article_text_preview {
  display: flex;
  margin: 0 auto;
  width: 100%;
  justify-content: center;
}

.article_attributes {
  font-size: 16px;
  font-weight: bold;
}

.article_attributes > span {
  margin-left: 3px;
}

.article_attributes > NavLink {
  color: black;
}
.article_attributes_username,
.article_attributes_category {
  text-decoration: none;
  color: rgb(51, 160, 255);
}

.article_attributes_username:hover,
.article_attributes_category:hover {
  text-decoration: underline;
}

.article_list_item_username:hover {
  cursor: pointer;
}

.container_vote_arrows {
  display: flex;
  flex-direction: row;
  justify-content: center;
  border-radius: 2px;
}

.lights_off_container_vote_arrows {
  border-right: 1px solid rgb(75, 74, 74);
}

.container_arrow_up {
  width: 50px;
  display: flex;
}

.arrow_up {
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-style: solid;
  border-color: black;
  border-width: 0px 3px 3px 0px;
  transform: rotate(225deg);
  transition: border-width 150ms ease-in-out;
  margin: 0 auto;
}

.voting_request_error {
  color: red;
  font-size: 12px;
  display: block;
  margin-bottom: 0px;
  padding-bottom: 20px;
  margin-top: 3px;
}

.container_vote_count {
  margin-bottom: 5px;
  justify-content: center;
  font-size: 16px;
}

.thumbs_up,
.thumbs_down {
  max-width: 16px;
}

.thumbs_up {
  margin-right: 8px;
}

.thumbs_down {
  margin-left: 8px;
}

.thumbs_up:hover,
.thumbs_down:hover {
  cursor: pointer;
}

.vote_count {
  margin: 0 auto;
  font-weight: bold;
}

.arrow_up:hover {
  border-bottom-width: 4px;
  border-right-width: 4px;
  cursor: pointer;
}

.container_arrow_down {
  display: flex;
  width: 50px;
}

.arrow_down {
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-style: solid;
  border-color: black;
  border-width: 0px 3px 3px 0px;
  transform: rotate(45deg);
  transition: border-width 150ms ease-in-out;
  margin: 0 auto;
}

.arrow_down:hover {
  border-bottom-width: 4px;
  border-right-width: 4px;
  cursor: pointer;
}

.article_list_item_lightsOff {
  background-color: rgb(209, 209, 209);
  color: black;
  border-top: 1px solid rgb(83, 82, 82);
  border-bottom: 1px solid rgb(83, 82, 82);
  box-shadow: 6px 6px 4px 2px rgb(15, 15, 15);
}

.lightsOff_text {
  color: black;
}

.lightsOff_arrow {
  border-color: black;
}

@media only screen and (max-width: 768px) {
  .container_article_list_item {
    padding-right: 0px;
  }

  .hide_image {
    display: none;
  }
}
