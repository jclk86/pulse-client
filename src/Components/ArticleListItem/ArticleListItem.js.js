import React, {Component} from "react"
import {NavLink, withRouter} from "react-router-dom"
// need to proptypes

class ArticleListItem extends Component {


  render() {
    const {article} = this.props
    return( // change the template string 
    <NavLink role="navigation" to={`/articles/:article_id`}>
      <div className="container_article_item">
        <h2 className="article_title">{article.title}</h2>
      </div>
    </NavLink>
    )
  }
}

export default withRouter(ArticleListItem)