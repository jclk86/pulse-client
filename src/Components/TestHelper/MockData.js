export const mockUser = {
  username: "Pop123",
  password: "Password123!",
  fullname: "Bob Smith",
  email: "BSmith@gmail.com"
};

export const mockArticle = {
  title: "test article",
  content: "test content",
  article_tag: "News",
  id: 1,
  date_created: "2019-10-14T21:49:20.570Z",
  author: {
    username: mockUser.username
  }
};

export const mockComment = {
  id: 1,
  content: "test comment content",
  date_created: "2019-10-14T21:49:20.570Z",
  article_id: 1,
  user: {
    username: mockArticle.author.username
  }
};
