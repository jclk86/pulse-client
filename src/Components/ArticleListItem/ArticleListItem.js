import React, { Component } from "react";
import { NavLink, withRouter } from "react-router-dom";
import { DateFormatter } from "../Utils/Utils";
import ThumbsUp from "../../images/thumbs-up.png";
import ThumbsDown from "../../images/thumbs-down.png";
import CommentApiService from "../../Services/comment-api-service";
import VoteApiService from "../../Services/vote-api-service";
import TokenService from "../../Services/token-service";
import ArticleListContext from "../../Context/ArticleListContext";
import PropTypes from "prop-types";
import "./ArticleListItem.css";

class ArticleListItem extends Component {
  static contextType = ArticleListContext;

  componentDidMount() {
    CommentApiService.getAllComments().then(this.context.setComments);
  }

  getTotalComments(article_id, comments) {
    return comments.filter(comment => article_id === comment.article_id).length;
  }

  filterTotalVotes(article_id, votes) {
    return votes.filter(vote => article_id === vote.article_id);
  }

  ellipsify(string) {
    if (string.length > 140) {
      return string.substring(0, 200) + "...";
    } else {
      return string;
    }
  }

  handleClickVoteUp(article_id) {
    this.context.clearError();
    VoteApiService.addVoteForArticle(article_id)
      .then(this.context.addVote)
      .catch(this.context.setError);
  }

  handleClickVoteDown(votes, article_id, user_id) {
    this.context.clearError();
    VoteApiService.deleteVote(article_id).catch(this.context.setError);
    this.context.deleteVote(votes, article_id, user_id);
  }

  render() {
    const { article } = this.props;
    const { comments, votes, error, lightsOff } = this.context;
    const numOfComments = this.getTotalComments(article.id, comments);
    const previewText = this.ellipsify(article.content);
    const totalVotes = this.filterTotalVotes(article.id, votes);
    const numOfVotes = totalVotes.length ? totalVotes.length : 0;
    const token = TokenService.readJwtToken();
    const user_id = token.user_id;
    return (
      <div className="container_article_list_item">
        <div
          className={`container_article_preview +  + ${
            lightsOff ? "" : "article_list_item_lightsOff"
          }`}
        >
          <NavLink to={`/articles/${article.id}`} role="navigation">
            <div className="container_article_list_item_image hide_image">
              <img
                src={`${article.image_url}`}
                alt={`${article.title}`}
                className="article_list_item_image"
              ></img>
            </div>
          </NavLink>
          <div className="container_article_text_preview">
            <div className="container_article_list_item_info">
              <NavLink
                role="navigation"
                to={`/articles/${article.id}`}
                className="article_list_item_title_link"
              >
                {" "}
                <h4
                  className={`article_list_item_title + ${
                    lightsOff ? "" : "lightsOff_text"
                  }`}
                >
                  {article.title}
                </h4>{" "}
              </NavLink>
              <p className="preview_text">{previewText}</p>
              <div className="container_article_attributes">
                <div className="article_attributes">
                  <div className="container_vote_count">
                    <p className="vote_count">
                      <img
                        src={ThumbsUp}
                        alt="vote up"
                        className="thumbs_up"
                        onClick={() => this.handleClickVoteUp(article.id)}
                      ></img>
                      {numOfVotes}
                      <img
                        src={ThumbsDown}
                        alt="vote up"
                        className="thumbs_down"
                        onClick={() =>
                          this.handleClickVoteDown(votes, article.id, user_id)
                        }
                      ></img>
                    </p>
                  </div>
                  <div className="voting_request_error">
                    {error && error.article_id === article.id && error.message}
                  </div>
                  <span>{numOfComments} comments</span>
                  <span>
                    /
                    <NavLink
                      to={`/articles/categories/${article.article_category}`}
                      className="article_attributes_category"
                    >
                      {article.article_category}
                    </NavLink>
                    /
                  </span>
                  <span>by</span>{" "}
                  <NavLink
                    to={`/profile/${article.author.username}`}
                    className={"article_attributes_username"}
                  >
                    {article.author.username}{" "}
                  </NavLink>
                  <span>on {DateFormatter(article.date_created)}</span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ArticleListItem.propTypes = {
  article: PropTypes.shape({
    image_url: PropTypes.string,
    title: PropTypes.string,
    content: PropTypes.string,
    date_created: PropTypes.string,
    author: PropTypes.shape({
      username: PropTypes.string
    })
  })
};

export default withRouter(ArticleListItem);

// <div
//   className={`container_vote_arrows + ${
//     lightsOff ? "" : "lights_off_container_vote_arrows"
//     }`}
// >
//   <div className="container_arrow_up">
//     <span className="voting_request_error">
//       {error && error.article_id === article.id && error.message}
//     </span>
//     <div
//       className={`arrow_up + ${lightsOff ? "" : "lightsOff_arrow"}`}
//       onClick={() => this.handleClickUpArrow(article.id)}
//     ></div>
//   </div>
//   <div className="container_vote_count">
//     <p className="vote_count">{numOfVotes}</p>
//   </div>
//   <div className="container_arrow_down">
//     <div
//       className={`arrow_down + ${
//         lightsOff ? "" : "lightsOff_arrow"
//         }`}
//       onClick={() =>
//         this.handleClickDownArrow(votes, article.id, user_id)
//       }
//     ></div>
//   </div>
// </div>
