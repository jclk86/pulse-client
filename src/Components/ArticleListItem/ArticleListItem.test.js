import React from "react";
import { BrowserRouter } from "react-router-dom";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import AuthApiService from "../../Services/auth-api-service";
import ArticleApiService from "../../Services/article-api-service";
import { mockUser, mockArticle } from "../TestHelper/MockData";
import ArticleListItem from "./ArticleListItem";

describe(`ArticleListItem Component`, () => {
  beforeAll(() => {
    AuthApiService.postUser(mockUser).then(user => {
      AuthApiService.postLogin({
        username: user.username,
        password: user.password
      });
    });

    ArticleApiService.postArticle(mockArticle);
  });

  it("renders ArticleListItem component without issues", () => {
    const wrapper = mount(
      <BrowserRouter>
        <ArticleListItem article={mockArticle}></ArticleListItem>
      </BrowserRouter>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
