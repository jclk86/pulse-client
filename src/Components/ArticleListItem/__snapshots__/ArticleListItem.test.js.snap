// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleListItem Component renders ArticleListItem component without issues 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(ArticleListItem)
      article={
        Object {
          "article_category": "News",
          "author": Object {
            "username": "Pop123",
          },
          "content": "test content",
          "date_created": "2019-10-14T21:49:20.570Z",
          "id": 1,
          "title": "test article",
        }
      }
    >
      <ArticleListItem
        article={
          Object {
            "article_category": "News",
            "author": Object {
              "username": "Pop123",
            },
            "content": "test content",
            "date_created": "2019-10-14T21:49:20.570Z",
            "id": 1,
            "title": "test article",
          }
        }
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      >
        <div
          className="container_article_list_item"
        >
          <div
            className="container_article_preview +  + article_list_item_lightsOff"
          >
            <NavLink
              role="navigation"
              to="/articles/1"
            >
              <Link
                aria-current={null}
                role="navigation"
                to={
                  Object {
                    "hash": "",
                    "pathname": "/articles/1",
                    "search": "",
                    "state": null,
                  }
                }
              >
                <LinkAnchor
                  aria-current={null}
                  href="/articles/1"
                  navigate={[Function]}
                  role="navigation"
                >
                  <a
                    aria-current={null}
                    href="/articles/1"
                    onClick={[Function]}
                    role="navigation"
                  >
                    <div
                      className="container_article_list_item_image hide_image"
                    >
                      <img
                        alt="test article"
                        className="article_list_item_image"
                        src="undefined"
                      />
                    </div>
                  </a>
                </LinkAnchor>
              </Link>
            </NavLink>
            <div
              className="container_article_text_preview"
            >
              <div
                className="container_article_list_item_info"
              >
                <NavLink
                  className="article_list_item_title_link"
                  role="navigation"
                  to="/articles/1"
                >
                  <Link
                    aria-current={null}
                    className="article_list_item_title_link"
                    role="navigation"
                    to={
                      Object {
                        "hash": "",
                        "pathname": "/articles/1",
                        "search": "",
                        "state": null,
                      }
                    }
                  >
                    <LinkAnchor
                      aria-current={null}
                      className="article_list_item_title_link"
                      href="/articles/1"
                      navigate={[Function]}
                      role="navigation"
                    >
                      <a
                        aria-current={null}
                        className="article_list_item_title_link"
                        href="/articles/1"
                        onClick={[Function]}
                        role="navigation"
                      >
                         
                        <h4
                          className="article_list_item_title + lightsOff_text"
                        >
                          test article
                        </h4>
                         
                      </a>
                    </LinkAnchor>
                  </Link>
                </NavLink>
                <p
                  className="preview_text"
                >
                  test content
                </p>
                <div
                  className="container_article_attributes"
                >
                  <div
                    className="article_attributes"
                  >
                    <div
                      className="container_vote_count"
                    >
                      <p
                        className="vote_count"
                      >
                        <img
                          alt="vote up"
                          className="thumbs_up"
                          onClick={[Function]}
                          src="thumbs-up.png"
                        />
                        0
                        <img
                          alt="vote up"
                          className="thumbs_down"
                          onClick={[Function]}
                          src="thumbs-down.png"
                        />
                      </p>
                    </div>
                    <div
                      className="voting_request_error"
                    />
                    <span>
                      0
                       comments
                    </span>
                    <span>
                      /
                      <NavLink
                        className="article_attributes_category"
                        to="/articles/categories/News"
                      >
                        <Link
                          aria-current={null}
                          className="article_attributes_category"
                          to={
                            Object {
                              "hash": "",
                              "pathname": "/articles/categories/News",
                              "search": "",
                              "state": null,
                            }
                          }
                        >
                          <LinkAnchor
                            aria-current={null}
                            className="article_attributes_category"
                            href="/articles/categories/News"
                            navigate={[Function]}
                          >
                            <a
                              aria-current={null}
                              className="article_attributes_category"
                              href="/articles/categories/News"
                              onClick={[Function]}
                            >
                              News
                            </a>
                          </LinkAnchor>
                        </Link>
                      </NavLink>
                      /
                    </span>
                    <span>
                      by
                    </span>
                     
                    <NavLink
                      className="article_attributes_username"
                      to="/profile/Pop123"
                    >
                      <Link
                        aria-current={null}
                        className="article_attributes_username"
                        to={
                          Object {
                            "hash": "",
                            "pathname": "/profile/Pop123",
                            "search": "",
                            "state": null,
                          }
                        }
                      >
                        <LinkAnchor
                          aria-current={null}
                          className="article_attributes_username"
                          href="/profile/Pop123"
                          navigate={[Function]}
                        >
                          <a
                            aria-current={null}
                            className="article_attributes_username"
                            href="/profile/Pop123"
                            onClick={[Function]}
                          >
                            Pop123
                             
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                    <span>
                      on 
                      October 14, 2019
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </ArticleListItem>
    </withRouter(ArticleListItem)>
  </Router>
</BrowserRouter>
`;
